packages/connect-core/src/transform-stream.ts

```diff
@@ -0,0 +1,320 @@
+// Copyright 2021-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { Code } from "./code.js";
+import { ConnectError } from "./connect-error.js";
+import type { EnvelopedMessage } from "./envelope.js";
+import type { Serialization } from "./serialization.js";
+import { compressedFlag, Compression } from "./compression.js";
+
+/**
+ * ValueOrEnd is either a value, or and end type. It is the counterpart of an
+ * enveloped message - which can have a flag indicating a special message at
+ * the end of the stream - but in deserialized form.
+ */
+type ValueOrEnd<T, E> = { end: false; value: T } | { end: true; value: E };
+
+/**
+ * Creates a WHATWG TransformStream that takes a value or special end type, and
+ * serializes it as an enveloped message.
+ *
+ * For example, an input with { end: true, value: ... } is serialized using
+ * the given endSerialization, and the resulting enveloped message has the
+ * given endStreamFlag.
+ *
+ * An input with { end: false, value: ... } is serialized using the given
+ * serialization, and the resulting enveloped message does not have the given
+ * endStreamFlag.
+ */
+export function transformSerializeWithEnd<T, E>(
+  serialization: Serialization<T>,
+  endSerialization: Serialization<E>,
+  endStreamFlag: number
+): TransformStream<ValueOrEnd<T, E>, EnvelopedMessage> {
+  return new TransformStream<ValueOrEnd<T, E>, EnvelopedMessage>({
+    transform(chunk, controller) {
+      let data: Uint8Array;
+      let flags = 0;
+      if (chunk.end) {
+        flags = flags | endStreamFlag;
+        try {
+          data = endSerialization.serialize(chunk.value);
+        } catch (e) {
+          return controller.error(
+            new ConnectError("failed to serialize end", Code.Internal)
```

Comment:
Should we specify `e` as the `cause` here (and a bunch of places below)?

Reply:
Yes, good shout. Updated.